igor-web/src/main/groovy/com/netflix/spinnaker/igor/wercker/WerckerService.groovy

```diff
@@ -0,0 +1,349 @@
+    }
+
+    String pipelineKey(Run run, Map<String, String> pipelineKeys) {
+        if (run.pipelineId) {
+            return pipelineKeys ? pipelineKeys.get(run.pipelineId) : getPipelineName(run.pipelineId);
+        } else {
+            return run.getApplication().owner.name + SPLITOR +
+                    run.getApplication().name + SPLITOR +
+                    run.getPipeline().name;
+        }
+    }
+
+    String getPipelineId(String appAndPipelineName) {
+        String[] split = appAndPipelineName.split(SPLITOR)
+        String owner = split[0]
+        String appName = split[1]
+        String pipelineName = split[2]
+        String pipelineId = cache.getPipelineID(groupKey, appAndPipelineName)
+        if (pipelineId == null) {
+            try {
+                List<Pipeline> pipelines = werckerClient.getPipelinesForApplication(authHeaderValue, owner, appName)
+                Pipeline matchingPipeline = pipelines.find {pipeline -> pipelineName == pipeline.name}
+                if (matchingPipeline) {
+                    pipelineId = matchingPipeline.id;
+                    cache.setPipelineID(groupKey, appAndPipelineName, pipelineId)
+                }
+            } catch(retrofit.RetrofitError err) {
+                log.info "Error getting pipelines for ${owner} ${appName} ${err} ${err.getClass()}"
+            }
+        }
+        return pipelineId;
+    }
+
+    String getPipelineName(String pipelineId) {
+        String name = cache.getPipelineName(groupKey, pipelineId)
+        if (name == null) {
+            try {
+                Pipeline pipeline = werckerClient.getPipeline(authHeaderValue, pipelineId)
+                if (pipeline && pipeline.application && pipeline.application.owner) {
+                    name = pipeline.application.owner.name + SPLITOR +
+                            pipeline.application.name  + SPLITOR +
+                            pipeline.name
+                    cache.setPipelineID(groupKey, name, pipelineId)
+                }
+            } catch(retrofit.RetrofitError err) {
+                log.info "Error getting pipelines for ${owner} ${appName} ${err} ${err.getClass()}"
+            }
+        }
+        return name;
+    }
+
+    Map<String, List<Run>> getRunsSince(Set<String> pipelines, long since) {
+        long start = System.currentTimeMillis();
+        List<String> pipelineIds = pipelines.collect { getPipelineId(it) };
+        Map<String, String> pipelineKeys = pipelines.collectEntries { [(getPipelineId(it)) : (it)] };
+        List<String> pids = pipelineKeys.keySet().asList();
+        Map<String, List<Run>> pipelineRuns = [:];
+        List<Run> allRuns = werckerClient.getRunsSince(authHeaderValue, branch, pids, limit, since);
+        log.debug "getRunsSince for pipelines:${pipelines} : ${allRuns.size()} runs in ${System.currentTimeMillis() - start}ms!!"
+        allRuns.forEach({ run ->
```

Comment:
Just a minor style point, but I think this block can be replaced with a Groovy `.groupBy`

Reply:
nice Groovyism suggestion - thx :) done